pkg/controller/deployment.go

```diff
@@ -0,0 +1,364 @@
+	"k8s.io/client-go/rest"
+
+	"sync"
+
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+)
+
+// DeploymentHandler interface contains the methods that are required
+type DeploymentHandler interface {
+	Added(ctx context.Context, obj *k8sAppsV1.Deployment) error
+	Deleted(ctx context.Context, obj *k8sAppsV1.Deployment) error
+}
+
+type DeploymentItem struct {
+	Deployment *k8sAppsV1.Deployment
+	Status     string
+}
+
+type DeploymentClusterEntry struct {
+	Identity    string
+	Deployments map[string]*DeploymentItem
+}
+
+type DeploymentController struct {
+	K8sClient         kubernetes.Interface
+	DeploymentHandler DeploymentHandler
+	Cache             *deploymentCache
+	informer          cache.SharedIndexInformer
+	labelSet          *common.LabelSet
+}
+
+type deploymentCache struct {
+	//map of dependencies key=identity value array of onboarded identities
+	cache map[string]*DeploymentClusterEntry
+	mutex *sync.Mutex
+}
+
+func NewDeploymentCache() *deploymentCache {
+	return &deploymentCache{
+		cache: make(map[string]*DeploymentClusterEntry),
+		mutex: &sync.Mutex{},
+	}
+}
+
+func (p *deploymentCache) getKey(deployment *k8sAppsV1.Deployment) string {
+	return common.GetDeploymentGlobalIdentifier(deployment)
+}
+
+func (p *deploymentCache) Get(key string, env string) *k8sAppsV1.Deployment {
+	defer p.mutex.Unlock()
+	p.mutex.Lock()
+
+	dce, ok := p.cache[key]
+	if ok {
+		dceEnv, ok := dce.Deployments[env]
+		if ok {
+			return dceEnv.Deployment
+		}
+	}
+
+	return nil
+}
+
+func (d *deploymentCache) List() []k8sAppsV1.Deployment {
+	var deploymentList []k8sAppsV1.Deployment
+	d.mutex.Lock()
+	defer d.mutex.Unlock()
+	for _, deploymentClusterEntry := range d.cache {
+		for _, deploymentItem := range deploymentClusterEntry.Deployments {
+			if deploymentItem != nil && deploymentItem.Deployment != nil {
+				deploymentList = append(deploymentList, *deploymentItem.Deployment)
+			}
+		}
+	}
+	return deploymentList
+}
+
+func (p *deploymentCache) GetDeploymentProcessStatus(deployment *k8sAppsV1.Deployment) string {
+	defer p.mutex.Unlock()
+	p.mutex.Lock()
+
+	env := common.GetEnv(deployment)
+	key := p.getKey(deployment)
+
+	dce, ok := p.cache[key]
+	if ok {
+		dceEnv, ok := dce.Deployments[env]
+		if ok {
+			return dceEnv.Status
+		}
+	}
+	return types.NotProcessed
+}
+
+func (p *deploymentCache) UpdateDeploymentProcessStatus(deployment *k8sAppsV1.Deployment, status string) error {
+	defer p.mutex.Unlock()
+	p.mutex.Lock()
+
+	env := common.GetEnv(deployment)
+	key := p.getKey(deployment)
+
+	dce, ok := p.cache[key]
+	if ok {
+		dceEnv, ok := dce.Deployments[env]
+		if ok {
+			dceEnv.Status = status
+			p.cache[dce.Identity] = dce
+			return nil
+		} else {
+			dce.Deployments[env] = &DeploymentItem{
+				Status: status,
+			}
+
+			p.cache[dce.Identity] = dce
+			return nil
+		}
+	}
+	return fmt.Errorf(types.LogCacheFormat, "Update", "Deployment",
```

Comment:
do we need env information also in the log?

Reply:
added